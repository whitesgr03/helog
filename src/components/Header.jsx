import style from "../styles/Header.module.css";
import { icon } from "../styles/image.module.css";
import { colorThemeToggle, sun, moon } from "../styles/Dropdown.module.css";

import { Link } from "react-router-dom";
import PropTypes from "prop-types";

import Dropdown from "./Dropdown";

const Header = ({ user }) => {
	// const darkTheme = useContext(themContext)
	// const user = useContext(userContext);
	const active = false;
	return (
		// <header className={`${darkTheme ? style.dark : ""} ${style.header} `}>
		<header className={style.header}>
			<Link to="/" className={style.logo}>
				<h1>HeLog</h1>
			</Link>
			<nav>
				<ul className={style.list}>
					{user.isAdmin && (
						<li>
							<Link to="/">
								<span className={`${icon} ${style.pencil}`} />
								Write
							</Link>
						</li>
					)}
					<li>
						<div className={style.switch}>
							<div className={colorThemeToggle}>
								{/* {darkTheme ? ( */}
								{/* <span className={`${icon} ${moon}`} /> */}
								{/* ) : ( */}
								<span className={`${icon} ${sun}`} />
								{/* )} */}
								<button>
									<div />
								</button>
							</div>
							<span>Light mode</span>
						</div>
					</li>
					<li>
						<button>
							<span className={`${icon} ${style.account}`} />
							Account
						</button>
					</li>
				</ul>
			</nav>
			{active && <Dropdown user={user} />}
		</header>
	);
};
Header.propTypes = {
	user: PropTypes.object,
};

export default Header;
