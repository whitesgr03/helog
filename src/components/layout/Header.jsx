// Packages
import { useState } from "react";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";

// Styles
import style from "../../styles/layout/Header.module.css";
import { sun, moon } from "../../styles/layout/Dropdown.module.css";
import button from "../../styles/utils/button.module.css";
import image from "../../styles/utils/image.module.css";
import { transparentWindow } from "../../styles/utils/bgc.module.css";

// Components
import Dropdown from "./Dropdown";

const Header = ({ user, darkTheme, handleSwitchColorTheme }) => {
	const [activeDropdown, setActiveDropdown] = useState(false);

	const handleActiveDropdown = () => setActiveDropdown(!activeDropdown);
	const handleCloseDropdown = () => setActiveDropdown(false);

	return (
		<>
			<header className={style.header}>
				<Link
					to="/"
					className={style.logo}
					onClick={handleCloseDropdown}
				>
					<h1>HeLog</h1>
				</Link>
				<nav>
					<ul className={style.list}>
						{user && (
							<li>
								<a href="https://helog-editor.localtest.me:7001">
									<span
										className={`${image.icon} ${style.pencil}`}
									/>
									Write
								</a>
							</li>
						)}
						<li className={style.toggleBtn}>
							<button onClick={handleSwitchColorTheme}>
								<div className={button.theme}>
									<span
										data-testid={"icon"}
										className={`${image.icon} ${
											darkTheme ? moon : sun
										}`}
									/>
									<div>
										<div />
									</div>
								</div>
								<span>{darkTheme ? "Dark" : "Light"} mode</span>
							</button>
						</li>
						<li>
							<button onClick={handleActiveDropdown}>
								<span
									className={`${image.icon} ${style.account}`}
								/>
								Account
							</button>
						</li>
					</ul>
				</nav>
				{activeDropdown && (
					<Dropdown
						user={user}
						darkTheme={darkTheme}
						handleSwitchColorTheme={handleSwitchColorTheme}
						handleCloseDropdown={handleCloseDropdown}
					/>
				)}
			</header>
			{activeDropdown && (
				<div
					className={transparentWindow}
					onClick={handleCloseDropdown}
					data-testid="transparentBgc"
				/>
			)}
		</>
	);
};
Header.propTypes = {
	user: PropTypes.object,
	darkTheme: PropTypes.bool,
	handleSwitchColorTheme: PropTypes.func,
};

export default Header;
