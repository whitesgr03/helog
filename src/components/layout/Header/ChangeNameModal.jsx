// Packages
import { useState, useRef, useEffect } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import PropTypes from 'prop-types';
import { string } from 'yup';
import isEmpty from 'lodash.isempty';
import { useMutation } from '@tanstack/react-query';
import { queryClient } from '../../../utils/queryOptions';

// Styles
import formStyles from '../../../styles/form.module.css';
import buttonStyles from '../../../styles/button.module.css';
import imageStyles from '../../../styles/image.module.css';

// Utils
import { updateUserInfo } from '../../../utils/handleUser';
import { verifySchema } from '../../../utils/verifySchema';

// Components
import { Loading } from '../../utils/Loading';

export const ChangeNameModal = ({ username, onActiveModal }) => {
	const [inputErrors, setInputErrors] = useState({});
	const [formFields, setFormFields] = useState({ username });
	const [debounce, setDebounce] = useState(false);

	const timer = useRef(null);
	const navigate = useNavigate();
	const { pathname: previousPath } = useLocation();

	const { isPending, mutate } = useMutation({
		mutationFn: updateUserInfo,
		onError: error =>
			navigate('/error', {
				state: { error: error.message, previousPath },
			}),
		onSuccess: data => {
			const handleSetUser = data => {
				queryClient.setQueryData(['userInfo'], data);
				onActiveModal({ component: null });
			};
			data.success ? handleSetUser(data) : setInputErrors({ ...data.fields });
		},
	});

	const handleSubmit = async e => {
		e.preventDefault();

		const schema = {
			username: string()
				.trim()
				.max(30, ({ max }) => `Username must be less than ${max} long.`)
				.matches(
					/^([a-zA-Z0-9](-|_|\s)?)*[a-zA-Z0-9]$/,
					'Username must be alphanumeric.',
				)
				.notOneOf(
					[username],
					'New username should be different from the old username.',
				)
				.required('Username is required.'),
		};

		const validationResult = await verifySchema({ schema, data: formFields });

		const handleInValid = () => {
			setInputErrors(validationResult.fields);
			setDebounce(false);
		};

		const handleValid = () => {
			setInputErrors({});
			mutate(formFields);
		};

		validationResult.success ? handleValid() : handleInValid();
	};

	const handleChange = e => {
		const { name, value } = e.target;
		const fields = {
			...formFields,
			[name]: value,
		};
		setFormFields(fields);
		!isEmpty(inputErrors) && setDebounce(true);
	};

	useEffect(() => {
		const schema = {
			username: string()
				.trim()
				.max(30, ({ max }) => `Username must be less than ${max} long.`)
				.matches(
					/^([a-zA-Z0-9](-|_|\s)?)*[a-zA-Z0-9]$/,
					'Username must be alphanumeric.',
				)
				.notOneOf(
					[username],
					'New username should be different from the old username.',
				)
				.required('Username is required.'),
		};
		debounce &&
			(timer.current = setTimeout(async () => {
				const validationResult = await verifySchema({
					schema,
					data: formFields,
				});
				validationResult.success
					? setInputErrors({})
					: setInputErrors(validationResult.fields);
			}, 500));

		return () => clearTimeout(timer.current);
	}, [debounce, username, formFields]);

	return (
		<>
			{isPending && <Loading text={'Saving...'} light={true} shadow={true} />}
			<form
				className={formStyles.content}
				onSubmit={e => !isPending && handleSubmit(e)}
			>
				<div className={formStyles['label-wrap']}>
					<label
						htmlFor="changeUserName"
						className={`${inputErrors?.username ? formStyles.error : ''}`}
					>
						Change username
						<input
							id="changeUserName"
							type="text"
							name="username"
							value={formFields.username}
							onChange={handleChange}
							spellCheck="false"
							autoCapitalize="off"
							autoCorrect="off"
							autoComplete="off"
						/>
					</label>
					<div>
						<span className={`${imageStyles.icon} ${formStyles.alert}`} />
						<span>{inputErrors.username ?? 'Message placeholder'}</span>
					</div>
				</div>

				<button
					type="submit"
					className={`${buttonStyles.content} ${buttonStyles.success}`}
				>
					Save
				</button>
			</form>
		</>
	);
};

ChangeNameModal.propTypes = {
	onActiveModal: PropTypes.func,
	username: PropTypes.string,
};
