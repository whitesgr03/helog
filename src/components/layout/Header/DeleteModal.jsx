// Packages
import { useState } from 'react';
import { useNavigate, useLocation } from 'react-router-dom';
import PropTypes from 'prop-types';

// Styles
import styles from './DeleteModal.module.css';
import buttonStyles from '../../../styles/button.module.css';

// Utils
import { deleteUser } from '../../../utils/handleUser';

// Components
import { Loading } from '../../utils/Loading';

export const DeleteModal = ({
	onUser,
	onToggleSettingsMenu,
	onActiveModal,
	onAlert,
}) => {
	const [loading, setLoading] = useState(false);
	const navigate = useNavigate();
	const { pathname: previousPath } = useLocation();

	const handleDelete = async () => {
		setLoading(true);

		const result = await deleteUser();

		const handleSetUser = () => {
			onUser(null);
			onAlert({
				message: 'Your account has been deleted.',
				error: false,
				delay: 2000,
			});
			onActiveModal({ component: null });
			onToggleSettingsMenu();
		};

		result.success
			? handleSetUser()
			: navigate('/error', {
					state: { error: result.message, previousPath },
				});

		setLoading(false);
	};
	return (
		<>
			{loading && <Loading text={'Deleting...'} light={true} shadow={true} />}
			<div className={styles.model}>
				<span className={styles.title}>Delete Your Account Forever</span>
				<span className={styles.content}>Do you really want to delete?</span>
				<div className={styles['button-wrap']}>
					<button
						className={`${buttonStyles.content} ${buttonStyles.cancel}`}
						data-close-model
						onClick={() => onActiveModal({ component: null })}
					>
						Cancel
					</button>
					<button
						className={`${buttonStyles.content} ${buttonStyles.error}`}
						onClick={handleDelete}
					>
						Delete
					</button>
				</div>
			</div>
		</>
	);
};

DeleteModal.propTypes = {
	onAlert: PropTypes.func,
	onUser: PropTypes.func,
	onToggleSettingsMenu: PropTypes.func,
	onActiveModal: PropTypes.func,
};
