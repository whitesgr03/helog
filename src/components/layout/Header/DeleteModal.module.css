.model {
	display: flex;
	row-gap: 20px;
	flex-direction: column;
	padding: 15px;
}

.title {
	font-weight: 700;
}

.content {
	align-self: center;
}

.button-wrap {
	display: flex;
	justify-content: space-evenly;
}

.button-wrap button {
	width: 40%;
}

@media (width >= 450px) {
	.model {
		row-gap: 40px;
		padding: 20px;
	}
}

@media (width >= 700px) {
	.model form {
		row-gap: 20px;
	}
}

@media (width >= 1000px) {
	.model {
		padding: 30px 40px;
	}
}
