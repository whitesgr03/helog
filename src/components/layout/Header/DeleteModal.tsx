// Packages
import { useNavigate, useLocation } from 'react-router-dom';
import { useMutation } from '@tanstack/react-query';
import { queryClient } from '../../../utils/queryOptions';

// Styles
import styles from './DeleteModal.module.css';
import buttonStyles from '../../../styles/button.module.css';

// Utils
import { deleteUser } from '../../../utils/handleUser';

// Components
import { Loading } from '../../utils/Loading';

// Context
import { useAppDataAPI } from '../../pages/App/AppContext';

interface DeleteModalProps {
	onCloseDropdown: () => void;
}

export const DeleteModal = ({ onCloseDropdown }: DeleteModalProps) => {
	const navigate = useNavigate();
	const { pathname: previousPath } = useLocation();
	const { onModal, onAlert } = useAppDataAPI();

	const { isPending, mutate } = useMutation({
		mutationFn: deleteUser,
		onError: () =>
			navigate('/error', {
				state: { previousPath },
			}),
		onSuccess: () => {
			queryClient.removeQueries({ queryKey: ['userInfo'] });
			onModal({ component: null });
			onCloseDropdown();
			onAlert([
				{
					message: 'Your account have been deleted.',
					error: false,
					delay: 4000,
				},
			]);
		},
	});

	const handleDelete = () => {
		onModal({
			component: <DeleteModal onCloseDropdown={onCloseDropdown} />,
			clickBgToClose: false,
		});
		mutate();
	};

	return (
		<>
			{isPending && <Loading text={'Deleting...'} light={true} shadow={true} />}
			<div className={styles.model}>
				<span className={styles.title}>Delete Your Account Forever</span>
				<span className={styles.content}>Do you really want to delete?</span>
				<div className={styles['button-wrap']}>
					<button
						className={`${buttonStyles.content} ${buttonStyles.cancel}`}
						data-close-model
						onClick={() => onModal({ component: null })}
					>
						Cancel
					</button>
					<button
						className={`${buttonStyles.content} ${buttonStyles.error}`}
						onClick={handleDelete}
					>
						Delete
					</button>
				</div>
			</div>
		</>
	);
};
