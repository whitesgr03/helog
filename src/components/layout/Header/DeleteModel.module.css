.model {
	display: flex;
	row-gap: 20px;
	flex-direction: column;
	padding: 15px;
}

.loading {
	display: flex;
	position: absolute;
	justify-content: center;
	align-items: center;
	z-index: 2;
	inset: 0;
	background-color: rgba(var(--dark-scheme-color-rgb), 0.8);
}

.title {
	font-weight: 700;
}

.content {
	align-self: center;
}

.button-wrap {
	display: flex;
	justify-content: space-evenly;
}

.button-wrap button {
	width: 40%;
}

/* Dark scheme classes */
:global(.dark) .loading {
	background-color: rgba(var(--light-scheme-color-rgb), 0.6);
}

@media (width >= 450px) {
	.model {
		row-gap: 40px;
		padding: 20px;
	}
}

@media (width >= 700px) {
	.model form {
		row-gap: 20px;
	}
}

@media (width >= 1000px) {
	.model {
		padding: 30px 40px;
	}
}
