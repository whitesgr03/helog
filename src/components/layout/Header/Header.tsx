// Packages
import { useState } from 'react';
import { Link } from 'react-router-dom';
import { useQuery } from '@tanstack/react-query';

// Styles
import styles from './Header.module.css';
import buttonStyles from '../../../styles/button.module.css';
import imageStyles from '../../../styles/image.module.css';

// Components
import { Dropdown } from './Dropdown';

// Type
import { DarkTheme } from '../../pages/App/App';

// Utils
import { queryUserInfoOption } from '../../../utils/queryOptions';

interface HeaderProps {
	darkTheme: DarkTheme;
	onColorTheme: () => void;
}

export const Header = ({ darkTheme, onColorTheme }: HeaderProps) => {
	const [activeDropdown, setActiveDropdown] = useState(false);

	const { data: user } = useQuery({ ...queryUserInfoOption(), enabled: false });

	const handleActiveDropdown = () => setActiveDropdown(!activeDropdown);
	const handleCloseDropdown = () => setActiveDropdown(false);

	return (
		<>
			<header className={styles.header}>
				<Link to="../" className={styles.logo} onClick={handleCloseDropdown}>
					<h1>HeLog</h1>
				</Link>
				<nav>
					<ul className={styles.list}>
						{user && (
							<li>
								<a
									href={`${import.meta.env.VITE_HELOG_EDITOR_URL}?theme=${darkTheme}`}
								>
									<span className={`${imageStyles.icon} ${styles.pencil}`} />
									Write
								</a>
							</li>
						)}
						<li className={styles['toggle-btn']}>
							<button onClick={onColorTheme}>
								<div className={buttonStyles.theme}>
									<span
										data-testid="theme-icon"
										className={`${imageStyles.icon} ${darkTheme ? styles.moon : styles.sun}`}
									/>
									<div>
										<div />
									</div>
								</div>
								{darkTheme ? 'Dark' : 'Light'} mode
							</button>
						</li>
						<li>
							<button onClick={handleActiveDropdown}>
								<span className={`${imageStyles.icon} ${styles.account}`} />
								Account
							</button>
						</li>
					</ul>
				</nav>
				{activeDropdown && (
					<>
						<Dropdown
							darkTheme={darkTheme}
							onColorTheme={onColorTheme}
							onCloseDropdown={handleCloseDropdown}
						/>
						<div
							className={styles['transparent-background']}
							onClick={handleCloseDropdown}
							data-testid="transparentBgc"
						/>
					</>
				)}
			</header>
		</>
	);
};
