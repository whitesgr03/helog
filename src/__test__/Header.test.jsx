import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";

import Header from "../components/Header";

describe("Header component", () => {
	it("should render correctly", () => {
		const { asFragment } = render(<Header />, {
			wrapper: BrowserRouter,
		});

		const actual = asFragment();

		expect(actual).toMatchSnapshot();
	});
});
