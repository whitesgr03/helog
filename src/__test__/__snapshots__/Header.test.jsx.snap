// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header component > should render correctly 1`] = `
<DocumentFragment>
  <header
    class="_header_a9d33c"
  >
    <h1>
      <a
        class="_logo_a9d33c"
        href="/"
      >
        HeLog
      </a>
    </h1>
    <nav>
      <ul
        class="_list_a9d33c"
      >
        <li>
          <a
            class="_link_a9d33c"
            href="/"
          >
            <span
              class="icon _pencil_a9d33c"
            />
            Write
          </a>
        </li>
        <li>
          <a
            class="_link_a9d33c"
            href="/users/login"
          >
            <span
              class="icon _account_a9d33c"
            />
            Account
          </a>
        </li>
      </ul>
    </nav>
  </header>
</DocumentFragment>
`;
